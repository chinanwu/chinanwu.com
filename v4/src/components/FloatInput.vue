<template>
  <div class="FloatInput">
    <label
      class="FloatInput__label"
      :class="{ 'FloatInput__label--float': float }"
      :id="'floatInputLabel-' + label"
      :for="'floatInput-' + label"
      :data-float="float"
    >
      {{ label }}
    </label>
    <span class="FloatInput__required" v-if="required">Required Field</span>
    <input
      :name="name"
      class="FloatInput__input"
      :id="'floatInput-' + label"
      :aria-labelledby="'floatInputLabel-' + label"
      type="text"
      v-model="inputValue"
      @focusin="handleFocusIn"
      @focusout="handleFocusOut"
      :accesskey="accessKey"
      :required="required"
    />
  </div>
</template>

<script>
export default {
  name: "FloatInput",
  props: {
    label: {
      type: String,
      required: true
    },
    name: {
      type: String,
      required: true
    },
    required: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      float: false,
      inputValue: ""
    };
  },
  emits: ["change"],
  methods: {
    handleFocusIn() {
      this.float = true;
    },
    handleFocusOut() {
      if (this.inputValue === "") {
        this.float = false;
      } else {
        this.$emit("change", this.inputValue);
      }
    }
  },
  computed: {
    accessKey() {
      return this.label[0].toLowerCase();
    }
  }
};
</script>

<style lang="less" scoped>
@import "../styling/colours.less";

.FloatInput {
  position: relative;
  margin-top: 3rem;
}

.FloatInput__label {
  position: absolute;
  font-size: 2.4rem;
  transition: all 0.2s;
  padding-left: 1rem;

  &[data-float="true"] {
    transform: translateY(-2rem);
    font-size: 1.6rem;
    padding-left: 0;
  }
}

.FloatInput__input {
  border: none;
  border-bottom: 1px solid black;
  background-color: @grey-06;
  font-size: 2.4rem;
  width: 100%;

  &[required] {
    box-shadow: none;
  }
}

.FloatInput__required {
  position: absolute;
  transform: translateY(4rem);
}
</style>
