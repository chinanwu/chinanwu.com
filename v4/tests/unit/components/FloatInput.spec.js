import { shallowMount } from "@vue/test-utils";
import FloatInput from "@/components/FloatInput.vue";

describe("FloatInput", () => {
  describe("rendering", () => {
    it("renders", () => {
      const wrapper = shallowMount(FloatInput, {
        props: {
          label: "label",
          name: "Test"
        }
      });
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders required field", () => {
      const wrapper = shallowMount(FloatInput, {
        props: {
          label: "label",
          name: "Test",
          required: true
        }
      });
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe("behaviour", () => {
    it("floats if focused in on", () => {
      const wrapper = shallowMount(FloatInput, {
        props: {
          label: "label",
          name: "Test"
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.vm.float).toBeFalsy();
      wrapper.find("#floatInput-label").trigger("focusin");
      expect(wrapper.vm.float).toBeTruthy();
    });

    it("stops floating if focus moves away with no input", () => {
      const wrapper = shallowMount(FloatInput, {
        props: {
          label: "label",
          name: "Test"
        },
        data() {
          return {
            float: true
          };
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      wrapper.find("#floatInput-label").trigger("focusout");
      expect(wrapper.vm.float).toBeFalsy();
    });

    it("keep floating if focus moves away with input", () => {
      const wrapper = shallowMount(FloatInput, {
        props: {
          label: "label",
          name: "Test"
        },
        data() {
          return {
            float: true,
            inputValue: "something"
          };
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      wrapper.find("#floatInput-label").trigger("focusout");
      expect(wrapper.vm.float).toBeTruthy();
      expect(wrapper.emitted().change).toEqual([["something"]]);
    });
  });
});
