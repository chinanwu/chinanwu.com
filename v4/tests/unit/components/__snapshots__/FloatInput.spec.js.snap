// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatInput behaviour floats if focused in on 1`] = `
<div
  class="FloatInput"
>
  <label
    class="FloatInput__label"
    data-float="false"
    for="floatInput-label"
    id="floatInputLabel-label"
  >
    label
  </label>
  <!--v-if-->
  <input
    accesskey="l"
    aria-labelledby="floatInputLabel-label"
    class="FloatInput__input"
    id="floatInput-label"
    name="Test"
    type="text"
  />
</div>
`;

exports[`FloatInput behaviour keep floating if focus moves away with input 1`] = `
<div
  class="FloatInput"
>
  <label
    class="FloatInput__label FloatInput__label--float"
    data-float="true"
    for="floatInput-label"
    id="floatInputLabel-label"
  >
    label
  </label>
  <!--v-if-->
  <input
    accesskey="l"
    aria-labelledby="floatInputLabel-label"
    class="FloatInput__input"
    id="floatInput-label"
    name="Test"
    type="text"
  />
</div>
`;

exports[`FloatInput behaviour stops floating if focus moves away with no input 1`] = `
<div
  class="FloatInput"
>
  <label
    class="FloatInput__label FloatInput__label--float"
    data-float="true"
    for="floatInput-label"
    id="floatInputLabel-label"
  >
    label
  </label>
  <!--v-if-->
  <input
    accesskey="l"
    aria-labelledby="floatInputLabel-label"
    class="FloatInput__input"
    id="floatInput-label"
    name="Test"
    type="text"
  />
</div>
`;

exports[`FloatInput rendering renders 1`] = `
<div
  class="FloatInput"
>
  <label
    class="FloatInput__label"
    data-float="false"
    for="floatInput-label"
    id="floatInputLabel-label"
  >
    label
  </label>
  <!--v-if-->
  <input
    accesskey="l"
    aria-labelledby="floatInputLabel-label"
    class="FloatInput__input"
    id="floatInput-label"
    name="Test"
    type="text"
  />
</div>
`;

exports[`FloatInput rendering renders required field 1`] = `
<div
  class="FloatInput"
>
  <label
    class="FloatInput__label"
    data-float="false"
    for="floatInput-label"
    id="floatInputLabel-label"
  >
    label
  </label>
  <span
    class="FloatInput__required"
  >
    Required Field
  </span>
  <input
    accesskey="l"
    aria-labelledby="floatInputLabel-label"
    class="FloatInput__input"
    id="floatInput-label"
    name="Test"
    required=""
    type="text"
  />
</div>
`;
